### PR TITLE
ci: `basic-workflow-tests.sh` improvments for 4xA10 workflow

### PR DESCRIPTION
**Pull request by alimaredia**
This PR has the following Features/PRs that put the 4xA10 workflow in an ideal state:
1.  Adding CACHE_DIR as a location to search for models: https://github.com/instructlab/instructlab/pull/1844
2.  Getting vLLM to work in basic-workflow.sh: https://github.com/instructlab/instructlab/pull/1841
The linked PR has multiple other fixes made to the test script, that are also in this PR. In the process of adding vLLM support, the `-g` flag (for GRANITE) was removed. Granite will now be the base model used in the entire workflow, mirroring the ideal ilab workflow.
3.  Adds remaining model evaluation benchmarks to the CI: https://github.com/instructlab/instructlab/pull/1676
4. Update ci.md to reflect the test matrix after this PR would be merged
5. Add the `-x` flag to `set` for faster debugging
6. Ensure evaluation runs when FULLTRAIN is not set

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

**Comment by alimaredia** [comment]
Run to ensure this branch passes A10x4 workflow: https://github.com/russellb/ilab-runner/actions/runs/10088144173/job/27893473117

File: .github/workflows/e2e-nvidia-a10g-x1.yml [hunk 1]
```diff
@@ -156,7 +156,7 @@ jobs:
 
           python3.11 -m pip show nvidia-nccl-cu12
           
-          ./scripts/basic-workflow-tests.sh -emf
+          ./scripts/basic-workflow-tests.sh -mf
```
Comment (alimaredia) [inline .github/workflows/e2e-nvidia-a10g-x1.yml:159]: The reason for this change was that MT-Bench-Branch uses merlinite-7b-lab as the base model. Merlinite-7b-lab cannot be served on a single A10 GPU with vLLM unless `--gpu-memory-utilization 1` is passed into vLLM. Currently the eval code doesn't have the ability to pass abitrary vLLM flags to individual models within MT-Bench to get merlinite to run.

CC @nathan-weinberg @alinaryan

File: docs/ci.md [hunk 2]
```diff
@@ -84,13 +84,13 @@ The project currently supports the usage of the following runners for the E2E jo
...
-|              | vllm                     |⎯|⎯|⎯|❌|
+|              | vllm                     |⎯|⎯|⎯|✅|
 | **Generate** | simple                   |✅|✅|✅|⎯|
 |              | full                     |⎯|⎯|⎯|✅|
-| **Training** | legacy+Linux             |⎯|⎯|✅|⎯|
+| **Training** | legacy+Linux             |⎯|⎯|❌|⎯|
 |              | legacy+Linux+4-bit-quant |✅|✅|⎯|⎯|
 |              | training-lib             |⎯|⎯|⎯|✅|
-| **Eval**     | eval                     |⎯|⎯|✅(*2)|❌️|
+| **Eval**     | eval                     |⎯|⎯|❌(*2)|✅|
```
Comment (alimaredia) [inline docs/ci.md:93]: The reasons for this change as mentioned above: https://github.com/instructlab/instructlab/pull/1860/files#r1690806223

File: docs/ci.md [hunk 3]
```diff
@@ -84,13 +84,13 @@ The project currently supports the usage of the following runners for the E2E jo
 | Area | Feature | [`e2e.yml`](https://github.com/instructlab/instructlab/blob/main/.github/workflows/e2e.yml) (small) ![`e2e.yaml` on `main`](https://github.com/instructlab/instructlab/actions/workflows/e2e.yml/badge.svg?branch=main)| [`e2e-nvidia-t4-x1.yml`](https://github.com/instructlab/instructlab/blob/main/.github/workflows/e2e-nvidia-t4-x1.yml) (small) | [`e2e-nvidia-a10g-x1.yml`](https://github.com/instructlab/instructlab/blob/main/.github/workflows/e2e-nvidia-a10g-x1.yml) (medium) | [`e2e-nvidia-a10g-x4.yml`](https://github.com/instructlab/instructlab/blob/main/.github/workflows/e2e-nvidia-a10g-x4.yml) (large) ![`e2e-nvidia-a10g-x4.yaml` on `main`](https://github.com/instructlab/instructlab/actions/workflows/e2e-nvidia-a10g-x4.yml/badge.svg?branch=main)|
 | --- | --- | --- | --- | --- | --- |
 | **Serving**  | llama-cpp                |✅|✅|✅|✅ (temporary)|
-|              | vllm                     |⎯|⎯|⎯|❌|
+|              | vllm                     |⎯|⎯|⎯|✅|
 | **Generate** | simple                   |✅|✅|✅|⎯|
 |              | full                     |⎯|⎯|⎯|✅|
-| **Training** | legacy+Linux             |⎯|⎯|✅|⎯|
+| **Training** | legacy+Linux             |⎯|⎯|❌|⎯|
```
Comment (alimaredia) [inline docs/ci.md:90]: This [run](https://github.com/instructlab/instructlab/actions/runs/10085764596/job/27887223060) reflects the reason for this change.

CC @RobotSail

Comment (RobotSail) [reply docs/ci.md:90]: @alimaredia You need to update the `--data-path` to point to the correct file path - since we run each end-to-end in a temporary directory, you have to use this directory when running the training.

**Comment by alimaredia** [comment]
Successful run here: https://github.com/russellb/ilab-runner/actions/runs/10090424876/job/27899918576

File: scripts/basic-workflow-tests.sh [hunk 4]
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env bash
-set -euf
+set -xeuf
```
Comment (leseb) [inline scripts/basic-workflow-tests.sh:2]: `-x` should probably go away once you're done testing?

File: docs/ci.md (hunk 3 above)
Comment (alimaredia) [reply docs/ci.md:89]: Thanks! That can get changed in another PR focused on getting the A10x1 workflow working, I just wanted to highlight the reason for the change

File: scripts/basic-workflow-tests.sh [hunk 5]
```diff
@@ -88,14 +93,9 @@ set_defaults() {
...
-    if [ "${GRANITE}" -eq 1 ] && [ "${MIXTRAL}" -eq 1 ]; then
-        echo "ERROR: Can not specify -g and -M at the same time."
-        exit 1
-    fi
-
-    if [ "${MIXTRAL}" -eq 1 ] && [ "${BACKEND}" = "vllm" ]; then
-        echo "ERROR: Can not specify -M and -v at the same time."
-        exit 1
+    if [ "${EVAL}" -eq 1 ] && [ "${BACKEND}" != "vllm" ]; then
+         echo "ERROR: Can must specify -e and -v at the same time."
```
Comment (leseb) [inline scripts/basic-workflow-tests.sh:97]: The message is quite confusing. Should it be: `ERROR: can not use eval if the backend is not vLLM`? Or did you mean `ERROR: must specify -e and -v at the same time.`?

File: scripts/basic-workflow-tests.sh [hunk 6]
```diff
@@ -210,13 +230,13 @@ test_taxonomy() {
 
 test_generate() {
     task Generate synthetic data
-    if [ "$GRANITE" -eq 1 ]; then
-        GENERATE_ARGS+=("--model" "${DATA_HOME}/instructlab/models/granite-7b-lab-Q4_K_M.gguf")
-    elif [ "$MIXTRAL" -eq 1 ]; then
-        GENERATE_ARGS+=("--model" "${DATA_HOME}/instructlab/models/mixtral-8x7b-instruct-v0.1.Q4_K_M.gguf")
-    elif [ "$BACKEND" = "vllm" ]; then
-        GENERATE_ARGS+=("--model" "${DATA_HOME}/instructlab/models/instructlab/merlinite-7b-lab")
+    GENERATE_ARGS+=("--endpoint-url" "http://localhost:8000/v1")
```
Comment (leseb) [inline scripts/basic-workflow-tests.sh:233]: Why is this needed? Isn't the code doing this by default already? I mean it should look into the `serve` section and find this default, right?

File: scripts/basic-workflow-tests.sh (hunk 5 above)
Comment (alimaredia) [reply scripts/basic-workflow-tests.sh:97]: Good catch, your suggestion for the error message is what I meant.

File: scripts/basic-workflow-tests.sh (hunk 6 above)
Comment (alimaredia) [reply scripts/basic-workflow-tests.sh:233]: If `--endpoint-url` is not specified, `ilab data generate` will start up a temp server in the background (https://github.com/instructlab/instructlab/blob/main/src/instructlab/data/generate.py#L196). 

In the future we might want this behavior in this test, but for now we're directly serving the teacher model in the test.

File: scripts/basic-workflow-tests.sh (hunk 4 above)
Comment (alimaredia) [reply scripts/basic-workflow-tests.sh:2]: The `-x` is incredibly useful for seeing the commands that actually get run and debugging. I couldn't have made this PR without it. I broke this out in a separate commit (https://github.com/instructlab/instructlab/pull/1860/commits/f9c13c3f19635ebd9eaad9a1f108990b06d5e588) for the sake of proposing the keep it permanently or cutting it from the PR entirely.

Since the log output from CI is usually only viewed when failures occur or for debugging, I think this change has quite a bit of benefit.

Comment (leseb) [reply scripts/basic-workflow-tests.sh:2]: I'm not opposed to having `-x`.

**Review by leseb — APPROVED** [review]

File: .github/workflows/e2e-nvidia-a10g-x1.yml (hunk 1 above)
Comment (nathan-weinberg) [reply .github/workflows/e2e-nvidia-a10g-x1.yml:159]: Can we just change the model to granite?

**Review by n1hility — APPROVED** [review]

File: scripts/basic-workflow-tests.sh [hunk 7]
```diff
@@ -301,43 +351,55 @@ test_exec() {
...
+    # so we skip trying to test the result
+    if [ "$FULLTRAIN" -eq 1 ]; then
+        # When you run this --
+        #   `ilab model convert` is only implemented for macOS with M-series chips for now
+        #test_convert
+    
+        test_serve trained "${DATA_HOME}/instructlab/checkpoints/model.gguf"
+        PID=$!
+
+       test_chat
```
Comment (booxter) [inline scripts/basic-workflow-tests.sh:365]: (No action required) Missing space in the indent!!! 🗡️

File: scripts/basic-workflow-tests.sh [hunk 8]
```diff
@@ -301,43 +351,55 @@ test_exec() {
...
     test_evaluate
 }
 
-wait_for_server(){
-    if ! timeout 120 bash -c '
-        until curl -sS http://localhost:8000/docs &> /dev/null; do
-            echo "waiting for server to start"
-            sleep 1
+wait_for_server() {
+    ACTION="$1"
```
Comment (booxter) [inline scripts/basic-workflow-tests.sh:377]: (Observation) The indent here is not very consistent. We should probably enforce some style with `bashate` or whatever the tool de jour.

File: scripts/basic-workflow-tests.sh [hunk 9]
```diff
@@ -301,43 +351,55 @@ test_exec() {
...
+    if [ "${ACTION}" == "start" ]; then
```
Comment (booxter) [inline scripts/basic-workflow-tests.sh:378]: (No action required) `==` is not POSIX compliant, it's usually better to use `=` which is identical and standard (but I understand the script is in `bash`.)

**Review by booxter — APPROVED** [review]
Thank you. (Leaving some nits or observations but they are not material.)
